### PR TITLE
Porter implementation of `/exec_work_order` (associated with TMapConKFrags for `/reencrypt` endpoint)

### LINKED ISSUE #2667: Scope Porter functionality needed by Bob
At a high-level Porter probably needs to support `Bob.retrieve`. Determine what underlying functionality needs to be provided to facilitate this workflow.

For Bob specifically:
* Learning about ursulas in treasure map
* Work Orders / capsule - can Porter do work order execution and attachment of cfrags to capsule without decryption or revealing sensitive information - and return capsule for Bob to do the decryption.
...others...?
 
Delivery here could be in the form of a comprehensive list of areas to be tackled, and a list of potential functions (and required inputs and outputs) that could be exposed by Porter. Extra credit for sequence diagrams 😅 j/k .

### PR DESCRIPTION
**Pull request by derekpierre**
**Type of PR:**
- [ ] Bugfix
- [X] Feature
- [ ] Documentation
- [ ] Other

**Required reviews:** 
- [ ] 1
- [X] 2
- [ ] 3

**What this does:**
POC implementation of `/exec_work_order` for Bob. The actual mechanics of the reencrypt operation is modified by TMapConKFrags since kFrags would not be stored by Ursula. 

Slight modification to original design of function - https://github.com/nucypher/nucypher/issues/2667#issuecomment-842416581 ('work_order' -> 'work_order_payload')

*Request*
```json
{
    "ursula": <checksum_address>,
    "work_order_payload": <bytes encoded as base64>
}
```
*Response*
```json
{
    "result": {
        "work_order_result": <bytes encoded as base64>
    }
}
```

**Issues fixed/closed:**
Related to #2704 
Closes #2667 

**Notes for reviewers:**
Based over #2733 .

`RestMiddleware.send_work_order_payload_to_ursula()` will be changed by TMapConKfrags to not require an associated arrangement id, so there is currently a stub for that function until TMapConKFrags gets merged. Therefore, actually trying to use the endpoint at the moment will not work properly.

### CODE REVIEW

**Review by KPrasch — COMMENTED** [review]
General high-level question:  What made you decide to go with `exec_work_order` instead of `retrieve`?

**Comment by derekpierre** [comment]
> General high-level question: What made you decide to go with exec_work_order instead of retrieve?

Traditionally "retrieve" was defined by `Bob.retrieve` (at least in the Python API/CLI) and it involved a series of steps including:
1. Getting the treasure map (probably no more with TMapConKFrags)
2. Processing the treasure map to obtain ursulas, and construct work orders
3. Sending work orders to ursula(s) to obtain cfrags
4. Attaching the returned cfrags to the capsule
5. Decrypting the data to obtain the cleartext

I viewed this as a single step in that process and did not want to confuse it with being the entire "retrieve" process.

One additional thought I had with this endpoint was the ability for `exec_work_order` to possibly generalize to any functionality, although I'm uncertain of the feasibility:

> This can potentially be extended for additional functionality under the guise of a generalized WorkOrder (perhaps?). It could be more of an Ursula do something you know how to do (DKG, TSS, etc.) with the relevant WorkOrder sub-type provided.

(See https://github.com/nucypher/nucypher/issues/2667#issuecomment-842416581)

**Comment by derekpierre** [comment]
Unsure why `test_policy_simple_sinpa` is failing - perhaps someone can point me in the right direction - perhaps I missed something with the original rebase. Same issue in `porter` branch - https://app.circleci.com/pipelines/github/nucypher/nucypher/8150/workflows/8130953c-39e3-4906-bca5-39d5ee05c720/jobs/138062.

**Review by piotr-roslaniec — APPROVED** [review]
LGTM :+1:

File: nucypher/utilities/porter/control/specifications/fields/workorder.py [hunk 1]
```diff
@@ -14,25 +14,15 @@
...
 
-from marshmallow import fields
-
-from nucypher.control.specifications.fields import BaseField
+from nucypher.control.specifications.fields import Base64BytesRepresentation
 from nucypher.policy.orders import WorkOrder as WorkOrderClass
 
 
-class WorkOrder(BaseField, fields.Field):
+class WorkOrder(Base64BytesRepresentation):
```
Comment (vepkenez) [inline nucypher/utilities/porter/control/specifications/fields/workorder.py:22]: I love this

**Review by vepkenez — APPROVED** [review]

File: docs/source/application_development/porter.rst [hunk 2]
```diff
@@ -490,12 +495,14 @@ Example Request
...
+        -d '{"treasure_map": "Qld7S8sbKFCv2B8KxfJo4oxiTOjZ4VPyqTK5K1xK6DND6TbLg2hvlGaMV69aiiC5QfadB82w/5q1Sw+SNFHN2e ...",
              "bob_encrypting_key": "026d1f4ce5b2474e0dae499d6737a8d987ed3c9ab1a55e00f57ad2d8e81fe9e9ac"}'
 
-.. note::
+OR
 
-    If URL parameters are used for the ``/publish_treasure_map`` endpoint instead of JSON data, the request will fail because the URL becomes too long.
+.. code:: bash
+
+    curl -X POST "<PORTER URI>/publish_treasure_map?treasure_map=Qld7S8sbKFCv2B8KxfJo4oxiTOjZ4VPyqTK5K1xK6DND6TbLg2hvlGaMV69aiiC5QfadB82w%2F5q1Sw%2BSNFHN2e ...&bob_encrypting_key=026d1f4ce5b2474e0dae499d6737a8d987ed3c9ab1a55e00f57ad2d8e81fe9e9ac"
```
Comment (derekpierre) [inline docs/source/application_development/porter.rst:505]: @piotr-roslaniec / @vepkenez - it turns out that I am able to execute `/publish_treasure_map` with URL parameters using the python requests library - the problem was the treasure_map value not being URL encoded 🤦 . Made a note about URL encoding in the docs, and that it is tool dependent.

**Pull request merged**